wc command

Nabegon opened the pull request:
お疲れ様です。wcコマンドの課題の提出になります。ご確認のほどよろしくお願いします。

Nabegon committed:
first commit

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+  end
+end
+
+def check_files(input)
```
filesと複数形になってますが、引数のファイルについてチェックしているだけなので単数形でいいのではないでしょうか

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+
+def show_help
+  puts "The path or option doesn't exist. Please type correct path or option."
+  exit
```
個別のメソッドでexitするのはびっくりするというか、呼び出し元のメイン処理部分でexitしたほうが見通しよい感じがしました。

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+  return if File.exist?(input)
+
+  puts "This file #{input} doesn't exist. Please type the correct file name"
+  exit
```
同上ですが、`check_files`というメソッド名からはputsしたりexitしたりがあまり想像出来ないです。97行目で
```
delta_l, delta_w, delta_c = get_file_nums(input, opt_l)
```
してますが、この前に普通に
```ruby
unless File.exist?(input)
  puts "This file #{input} doesn't exist. Please type the correct file name"
  exit
end
```
ってある方が見通しいい感じがしました。`get_file_nums`の中でこのメソッド読んでますが、それもメソッド名から内部でやってることが想像つきにくいなと感じます。

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+end
+
+def get_file_nums(input, opt_l)
+  check_files(input)
```
上で書いた通りですがここでやらなくていいんじゃないか、と感じます。

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+    w_num += words.size
+  end
+
+  print_with_filename(opt_l, l_num, w_num, c_num, input)
```
`get_file_nums`というメソッド名から出力するとは思わないので、出力は呼び出しもとで戻り値受け取ってからやるほうがリーダブルなコードかなーと思いました

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+  [l_num, w_num, c_num]
+end
+
+def load_all_data(opt_l, str)
```
loadっていうよりやってるとは出力ですよねー。メソッド名に違和感あるなと思いました。

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+  print_all(opt_l, lines, words, bytes)
+end
+
+def load_lines(str)
```
loadってより値の算出をしている感じじゃないですかね。別の動詞がわかりやすいように感じました。いかがでしょう?

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+
+def load_words(str)
+  str.split(/\s+/).size
+end
```
同上です

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+
+def load_bytes(str)
+  str.size
+end
```
同上です

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+  end
+end
+
+error = OptionParser::InvalidOption
```
複数ヶ所で使うわけでもないですし、変数におく必要ないかなと思いました

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+sum_l = sum_w = sum_c = 0
+delta_l = delta_w = delta_c = 0
+
+if !file_name.nil?
```
変数におくまでもなく、`if ARGV[0]`で判定していいかなーと思いました。あえて否定にするのもややこしくなっているだけかなと。

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,106 @@
+    sum_w += delta_w
+    sum_c += delta_c
+  end
+  print_sum_total(opt_l, sum_l, sum_w, sum_c) if sum_l != delta_l
```
> コマンドライン引数の場合、引数ごとにループを回しているので、totalの計算の部分が無理矢理っぽくなってしまいました。こうした複数入力がある場合、最初に入力を配列にまとめて、その配列をそれぞれのメソッドに渡していく方が良いでしょうか。

ここのことですよね? そんなに気にならないですよ。

> 最初に入力を配列にまとめて、その配列をそれぞれのメソッドに渡していく方が良いでしょうか。

こうしてもいいかとは思いますが、このままでもいいんじゃないでしょうか。

Nabegon committed:
modified after 1st review

Nabegon committed:
add main method, change the order of methods, modify the inside of check_file() method

Nabegon committed:
remove: unused method

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,94 @@
+
+def main
+  options = ARGV.getopts('l')
+  options.nil? ? show_help && exit(1) : option_l = options['l']
```
せっかくmain関数導入したので、exitはもはや使わずに普通にreturnする方がいいかなと思いました。`早期return`ってやつですね。

動作確認してませんが、
```ruby
  if optoins.empty?
    show_help
    return
  end

  option_l = options['l']
```
みたいな? すると`else`の分岐一つ減るかなと思います。いかがでしょうか。

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,94 @@
+  if !ARGV[0].nil?
+    ARGV.each_with_index do |_arg, i|
+      input = ARGV[i]
+      exit(1) unless check_file(input)
```
ここも`return`でいいんじゃないですかね。`exit`って強制終了ですし、使わずに済むなら避けたいかなと。

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,94 @@
+  options.nil? ? show_help && exit(1) : option_l = options['l']
+  sum_l = sum_w = sum_c = delta_l = delta_w = delta_c = 0
+
+  if !ARGV[0].nil?
```
前回指摘したような気もしつつ、なるべく肯定条件で書いたほうが読みやすいので
```ruby
if ARGV[0]
```
で良かったりしませんか?

maedana commented on 06.wc/wc.rb:
```diff
@@ -0,0 +1,94 @@
+  else
+    puts "#{l_num} #{w_num} #{c_num} #{input}"
+  end
+end
```
> if文を1行で書けませんでした。1行で書く場合、どのような変更が可能でしょうか。ご教示いただきたいです。

`option_l ? puts puts "#{sum_l} total" : puts "#{sum_l} #{sum_w} #{sum_c} total"` って書いたんですよね?

```
irb(main):001:0' true ? puts 'true' : puts 'false'
Traceback (most recent call last):
        3: from /home/maedana/.rbenv/versions/3.0.0/bin/irb:23:in `<main>'
        2: from /home/maedana/.rbenv/versions/3.0.0/bin/irb:23:in `load'
        1: from /home/maedana/.rbenv/versions/3.0.0/lib/ruby/gems/3.0.0/gems/irb-1.3.0/exe/irb:11:in `<top (required)>'
SyntaxError ((irb):1: syntax error, unexpected string literal, expecting `do' or '{' or '(')
true ? puts 'true' : puts 'false'
            ^
(irb):1: syntax error, unexpected ':', expecting end-of-input
true ? puts 'true' : puts 'false'
                   ^
irb(main):002:0> true ? puts('true') : puts('false')
true
=> nil
```
なので、syntax errorだったんじゃないですかね? 括弧つけたらたぶんいけるんじゃないですかね?さらにいうとputsは外に出して
```
irb(main):003:0> puts (true ? 'true' : 'false')
true
```

って書き方でいいんじゃないか、とは思います。Rubyのifは式で値を返すので、putsに渡す値を三項演算子で返すイメージです。

maedana replied:
> option lの有無で処理を別けています。こちら、他に良い書き方があったら教えていただきたいです。

これでいいんじゃないかなーと思いました。

Nabegon committed:
modify

instead of using exit(1), return and break are used. For that, there is a new method check_option and rescue exception.

Nabegon committed:
modify: !ARGV[0].nil? -> ARGV[0]

Nabegon committed:
modify

in print_with_filename() and print_sum_total(), ternary operator is used instead of if statement; the value to be passed to puts is returned by the ternary operator.
